src/Nethermind/Nethermind.Core/Extensions/Bytes.Vector.cs

```diff
@@ -0,0 +1,129 @@
+
+using System;
+using System.Buffers;
+using System.Runtime.CompilerServices;
+using System.Runtime.InteropServices;
+using System.Runtime.Intrinsics;
+using System.Runtime.Intrinsics.X86;
+
+namespace Nethermind.Core.Extensions
+{
+    public static unsafe partial class Bytes
+    {
+        private static readonly byte[] ReverseMask = {15, 14, 13, 12, 11, 10, 9, 8, 7, 6, 5, 4, 3, 2, 1, 0, 15, 14, 13, 12, 11, 10, 9, 8, 7, 6, 5, 4, 3, 2, 1, 0};
+        private static readonly Vector256<byte> ReverseMaskVec;
+
+        static Bytes()
+        {
+            if (Avx2.IsSupported)
+            {
+                unsafe
+                {
+                    fixed (byte* ptr_mask = ReverseMask)
+                    {
+                        ReverseMaskVec = Avx2.LoadVector256(ptr_mask);
+                    }
+                }
+            }
+        }
+
+        public static void Avx2Reverse256InPlace(Span<byte> bytes)
+        {
+            unsafe
+            {
+                fixed (byte* inputPointer = bytes)
+                {
+                    Vector256<byte> inputVector = Avx2.LoadVector256(inputPointer);
+                    Vector256<byte> resultVector = Avx2.Shuffle(inputVector, ReverseMaskVec);
+                    resultVector = Avx2.Permute4x64(resultVector.As<byte, ulong>(), 0b01001110).As<ulong, byte>();
+
+                    Avx2.Store(inputPointer, resultVector);
+                }
+            }
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveOptimization)]
+        public static void Or(this Span<byte> thisSpam, Span<byte> valueSpam)
+        {
+            var length = thisSpam.Length;
+            if (length != valueSpam.Length)
+            {
+                throw new ArgumentException("Both byte spans has to be same length.");
+            }
+
+            int i = 0;
+
+            fixed (byte* thisPtr = thisSpam)
+            fixed (byte* valuePtr = valueSpam)
+            {
+                if (Avx2.IsSupported)
+                {
+                    for (; i < length - (Vector256<byte>.Count - 1); i += Vector256<byte>.Count)
+                    {
+                        Vector256<byte> b1 = Avx2.LoadVector256(thisPtr + i);
+                        Vector256<byte> b2 = Avx2.LoadVector256(valuePtr + i);
+                        Avx2.Store(thisPtr + i, Avx2.Or(b1, b2));
+                    }
+                }
+                else if (Sse2.IsSupported)
+                {
+                    for (; i < length - (Vector128<byte>.Count - 1); i += Vector128<byte>.Count)
+                    {
+                        Vector128<byte> b1 = Sse2.LoadVector128(thisPtr + i);
+                        Vector128<byte> b2 = Sse2.LoadVector128(valuePtr + i);
+                        Sse2.Store(thisPtr + i, Sse2.Or(b1, b2));
+                    }
+                }
+            }
+
+            for (; i < length; i++)
+            {
+                thisSpam[i] |= valueSpam[i];
```

Comment:
ref var refA = ref MemoryMarshal.AsRef<ulong>(a);
                            ref var refB = ref MemoryMarshal.AsRef<ulong>(b);
                            ref var refBuffer = ref MemoryMarshal.AsRef<ulong>(stack.Register);

                            refBuffer = refA | refB;
                            Unsafe.Add(ref refBuffer, 1) = Unsafe.Add(ref refA, 1) | Unsafe.Add(ref refB, 1);
                            Unsafe.Add(ref refBuffer, 2) = Unsafe.Add(ref refA, 2) | Unsafe.Add(ref refB, 2);
                            Unsafe.Add(ref refBuffer, 3) = Unsafe.Add(ref refA, 3) | Unsafe.Add(ref refB, 3);
...

Reply:
this ors 8 bytes at a time
so will be 64 times faster or more

Reply:
you also remove the loop

Reply:
I suggest writing specifically an optimized Bloom or operation instead of generalized Bytes

Reply:
Bloom being 256 bytes never goes into this loop if there is AVX/SSE2. This loop is only for leftovers of earlier AVX/SSE2 processing or for devices with no AVX/SSE2. I don't understand your code so lets have a chat about it.